igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,289 @@
+          error if something goes wrong
+        - `None`: try and discover a valid cert, but
+          try something else if that fails
+
+    token : `scitokens.SciToken`, `str`, `bool`, optional
+        bearer token (scitoken) input, one of
+
+        - an authorisation bearer token (`scitokens.SciToken`),
+        - a serialised token (`str`, `bytes`),
+        - `False`: disable using tokens completely
+        - `True`: discover a valid token via
+          :func:`igwn_auth_utils.find_scitoken` and
+          error if something goes wrong
+        - `None`: try and discover a valid token, but
+          try something else if that fails
+
+    token_audience, token_scope : `str`
+        the ``audience`` and ``scope`` to pass to
+        `igwn_auth_utils.find_scitoken` when discovering available tokens.
+
+    force_noauth : `bool`, optional
+        disable the use of any authorisation credentials (mainly for testing).
+
+    fail_if_noauth : `bool`, optional
+        raise an `Exception` if no authorisation credentials are presented
+        or discovered.
+
+    Raises
+    ------
+    ValueError
+        if ``fail_if_noauth=True`` is given and no authorisation
+        token/credentials are presented or discovered.
+
+    RuntimeError
+        if ``cert=True`` or ``token=True`` is given and the relevant
+        credential was not actually discovered.
+
+    Note
+    ----
+    This class requires `Requests <https://requests.readthedocs.io/>`__.
+
+    See also
+    --------
+    requests.Session
+        for details of the standard options
+    """.strip()
+
+
+class SessionAuthMixin:
+    """Mixin for :class:`requests.Session` to add support for IGWN auth.
+
+    By default this mixin will automatically attempt to discover/configure
+    a bearer token (scitoken) or an X.509 credential, with options to
+    require/disable either of those, or all authentication entirely.
+
+    {parameters}
+    """
+    def __init__(
+        self,
+        auth=None,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        url=None,
+        force_noauth=False,
+        fail_if_noauth=False,
+        **kwargs,
+    ):
+        # initialise session
+        super().__init__(**kwargs)
+
+        # handle options
+        if force_noauth and fail_if_noauth:
+            raise ValueError(
+                "cannot select both force_noauth and fail_if_noauth",
+            )
+        if force_noauth:
+            return
+
+        # find creds if we can
+        self._init_auth(
+            auth=auth,
+            cert=cert,
+            token=token,
+            token_audience=token_audience,
+            token_scope=token_scope,
+            url=url,
+        )
+
+        # if no auth was found, and we need it, fail here
+        if not self._has_igwn_auth and fail_if_noauth:
+            raise ValueError("no valid authorisation credentials found")
```

Comment:
Consider declaring a new exception class so that client code can catch this error explicitly. Take a look at existing Requests exception classes to see if there is one that would make sense as the base class.

Reply:
Done in latest commit, see the new `IgwnAuthError` object.